File: src/main/java/dark/leech/text/plugin/js/api/Regexp.java
package dark.leech.text.plugin.js.api;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.regex.PatternSyntaxException;

import org.mozilla.javascript.Context;
import org.mozilla.javascript.Scriptable;

import dark.leech.text.action.Log;

/**
 * Regular Expression API for JavaScript plugins using Rhino. Provides pattern matching, finding,
 * replacing, and splitting.
 */
public class Regexp extends JsApiWrapper {

    /**
     * Create Regexp API with execution context for proper object creation.
     *
     * @param context The Rhino Context for script execution
     * @param scope The Rhino Scriptable scope
     */
    public Regexp(Context context, Scriptable scope) {
        super(context, scope);
    }

    /** Create Regexp API without context (legacy compatibility). */
    public Regexp() {
        super();
    }

    /** Find first match in text. Usage: regexp.find("Hello World", "llo") => "llo" */
    public String find(String text, String pattern) {
        if (text == null || pattern == null) {
            return "";
        }
        try {
            Pattern p = Pattern.compile(pattern);
            Matcher m = p.matcher(text);
            if (m.find()) {
                return m.group();
            }
            return "";
        } catch (PatternSyntaxException e) {
            Log.add("Invalid regex pattern: " + pattern);
            return "";
        }
    }

    /** Find all matches in text. Usage: regexp.findAll("Hello World", "l") => ["l", "l", "l"] */
    public Object[] findAll(String text, String pattern) {
        List<String> results = new ArrayList<>();
        if (text != null && pattern != null) {
            try {
                Pattern p = Pattern.compile(pattern);
                Matcher m = p.matcher(text);
                while (m.find()) {
                    results.add(m.group());
                }
            } catch (PatternSyntaxException e) {
                Log.add("Invalid regex pattern: " + pattern);
            }
        }
        return results.toArray();
    }

    /** Match all with groups and indices. Usage: regexp.matchAll("test123", "\\d+") */
    public Object[] matchAll(String text, String pattern) {
        List<Map<String, Object>> results = new ArrayList<>();
        if (text == null || pattern == null) {
            return new Object[0];
        }

        try {
            Pattern p = Pattern.compile(pattern);
            Matcher m = p.matcher(text);

            int matchCount = 0;
            while (m.find()) {
                Map<String, Object> matchInfo = new HashMap<>();
                matchInfo.put("match", m.group());
                matchInfo.put("start", m.start());
                matchInfo.put("end", m.end());
                matchInfo.put("index", matchCount++);

                // Add groups
                int groupCount = m.groupCount();
                if (groupCount > 0) {
                    List<String> groups = new ArrayList<>();
                    for (int i = 1; i <= groupCount; i++) {
                        groups.add(m.group(i));
                    }
                    matchInfo.put("groups", groups.toArray());
                }

                results.add(matchInfo);
            }
        } catch (PatternSyntaxException e) {
            Log.add("Invalid regex pattern: " + pattern);
        }

        return results.toArray();
    }

    /** Replace first occurrence. */
    public String replace(String text, String pattern, String replacement) {
        if (text == null || pattern == null) {
            return text != null ? text : "";
        }
        try {
            Pattern p = Pattern.compile(pattern);
            Matcher m = p.matcher(text);
            if (m.find()) {
                return m.replaceFirst(replacement != null ? replacement : "");
            }
            return text;
        } catch (PatternSyntaxException e) {
            Log.add("Invalid regex pattern: " + pattern);
            return text;
        }
    }

    /** Replace all occurrences. */
    public String replaceAll(String text, String pattern, String replacement) {
        if (text == null || pattern == null) {
            return text != null ? text : "";
        }
        try {
            Pattern p = Pattern.compile(pattern);
            return p.matcher(text).replaceAll(replacement != null ? replacement : "");
        } catch (PatternSyntaxException e) {
            Log.add("Invalid regex pattern: " + pattern);
            return text;
        }
    }

    /** Split text by pattern. */
    public Object[] split(String text, String pattern) {
        if (text == null || pattern == null) {
            return text != null ? new Object[] {text} : new Object[0];
        }
        try {
            return text.split(pattern, -1);
        } catch (PatternSyntaxException e) {
            Log.add("Invalid regex pattern: " + pattern);
            return new Object[] {text};
        }
    }

    /** Test if pattern matches. */
    public boolean test(String text, String pattern) {
        if (text == null || pattern == null) {
            return false;
        }
        try {
            Pattern p = Pattern.compile(pattern);
            return p.matcher(text).find();
        } catch (PatternSyntaxException e) {
            Log.add("Invalid regex pattern: " + pattern);
            return false;
        }
    }

    /** Escape regex special characters. */
    public String escape(String text) {
        if (text == null) {
            return "";
        }
        return Pattern.quote(text);
    }

    /** Find last match. */
    public String findLast(String text, String pattern) {
        if (text == null || pattern == null) {
            return "";
        }
        try {
            Pattern p = Pattern.compile(pattern);
            Matcher m = p.matcher(text);
            String lastMatch = "";
            while (m.find()) {
                lastMatch = m.group();
            }
            return lastMatch;
        } catch (PatternSyntaxException e) {
            Log.add("Invalid regex pattern: " + pattern);
            return "";
        }
    }

    /** Count occurrences. */
    public int count(String text, String pattern) {
        if (text == null || pattern == null) {
            return 0;
        }
        int count = 0;
        try {
            Pattern p = Pattern.compile(pattern);
            Matcher m = p.matcher(text);
            while (m.find()) {
                count++;
            }
        } catch (PatternSyntaxException e) {
            Log.add("Invalid regex pattern: " + pattern);
        }
        return count;
    }

    /** Get all groups from matches. */
    public Object[] groups(String text, String pattern) {
        List<String> groupList = new ArrayList<>();
        if (text == null || pattern == null) {
            return new Object[0];
        }

        try {
            Pattern p = Pattern.compile(pattern);
            Matcher m = p.matcher(text);

            if (m.find()) {
                int groupCount = m.groupCount();
                for (int i = 0; i <= groupCount; i++) {
                    groupList.add(m.group(i));
                }
            }
        } catch (PatternSyntaxException e) {
            Log.add("Invalid regex pattern: " + pattern);
        }

        return groupList.toArray();
    }

    /** Extract named groups (if pattern uses named groups). */
    public Map<String, String> namedGroups(String text, String pattern) {
        Map<String, String> result = new HashMap<>();
        if (text == null || pattern == null) {
            return result;
        }

        try {
            Pattern p = Pattern.compile(pattern);
            Matcher m = p.matcher(text);

            if (m.find()) {
                // Try to extract named groups using reflection
                // Note: Java regex doesn't directly support named groups extraction
                // This is a simplified implementation
                int groupCount = m.groupCount();
                for (int i = 1; i <= groupCount; i++) {
                    result.put("group" + i, m.group(i));
                }
            }
        } catch (PatternSyntaxException e) {
            Log.add("Invalid regex pattern: " + pattern);
        }

        return result;
    }

    /** Check if pattern is valid. */
    public boolean isValid(String pattern) {
        if (pattern == null || pattern.isEmpty()) {
            return false;
        }
        try {
            Pattern.compile(pattern);
            return true;
        } catch (PatternSyntaxException e) {
            return false;
        }
    }

    /** Get match at specific index. */
    public String matchAt(String text, String pattern, int index) {
        if (text == null || pattern == null || index < 0) {
            return "";
        }

        try {
            Pattern p = Pattern.compile(pattern);
            Matcher m = p.matcher(text);

            int currentIndex = 0;
            while (m.find()) {
                if (currentIndex == index) {
                    return m.group();
                }
                currentIndex++;
            }
        } catch (PatternSyntaxException e) {
            Log.add("Invalid regex pattern: " + pattern);
        }

        return "";
    }

    /** Get all matches with their positions. */
    public Object[] matchesWithPositions(String text, String pattern) {
        List<Map<String, Object>> results = new ArrayList<>();
        if (text == null || pattern == null) {
            return new Object[0];
        }

        try {
            Pattern p = Pattern.compile(pattern);
            Matcher m = p.matcher(text);

            while (m.find()) {
                Map<String, Object> match = new HashMap<>();
                match.put("text", m.group());
                match.put("start", m.start());
                match.put("end", m.end());
                results.add(match);
            }
        } catch (PatternSyntaxException e) {
            Log.add("Invalid regex pattern: " + pattern);
        }

        return results.toArray();
    }
}

File: engine/src/commonMain/kotlin/dev/haipham22/leechtext/plugin/js/api/Regexp.kt
package dev.haipham22.leechtext.plugin.js.api

import dev.haipham22.leechtext.log.EngineLogger

/** Regex API cho JS plugins (port từ Regexp.java; P5.2c: kotlin.text.Regex thay
 * java.util.regex — semantic match: find/findAll/replace/split theo pattern). */
@Suppress("TooManyFunctions") // API surface cho plugin JS — số hàm mirror đúng API vBook, không tách
class Regexp(
    private val log: EngineLogger,
) {
    /** Compile pattern; log + null khi pattern invalid (giống bản gốc). */
    private fun compiled(pattern: String): Regex? = try {
        Regex(pattern)
    } catch (e: Exception) {
        log.add("Invalid regex pattern: $pattern")
        null
    }

    /** Find match đầu tiên. Usage: regexp.find("Hello World", "llo") => "llo" */
    fun find(
        text: String?,
        pattern: String?,
    ): String {
        if (text == null || pattern == null) return ""
        val m = compiled(pattern) ?: return ""
        return m.find(text)?.value ?: ""
    }

    /** Find tất cả matches. Usage: regexp.findAll("Hello World", "l") => ["l", "l", "l"] */
    fun findAll(
        text: String?,
        pattern: String?,
    ): Array<String> {
        val results = ArrayList<String>()
        if (text != null && pattern != null) {
            val m = compiled(pattern)
            if (m != null) m.findAll(text).forEach { results.add(it.value) }
        }
        return results.toTypedArray()
    }

    /** Match tất cả kèm groups và indices. Usage: regexp.matchAll("test123", "\\d+") */
    fun matchAll(
        text: String?,
        pattern: String?,
    ): Array<HashMap<String, Any?>> {
        if (text == null || pattern == null) return arrayOf()

        val results = ArrayList<HashMap<String, Any?>>()
        val regex = compiled(pattern) ?: return arrayOf()
        var matchCount = 0
        for (match in regex.findAll(text)) {
            val matchInfo = HashMap<String, Any?>()
            matchInfo["match"] = match.value
            matchInfo["start"] = match.range.first
            matchInfo["end"] = match.range.last + 1
            matchInfo["index"] = matchCount++

            if (match.groups.size > 1) {
                val groups = ArrayList<String?>()
                for (i in 1 until match.groups.size) groups.add(match.groups[i]?.value)
                matchInfo["groups"] = groups.toTypedArray()
            }

            results.add(matchInfo)
        }
        return results.toTypedArray()
    }

    /** Replace match đầu tiên theo pattern (tên gốc vBook — alias replaceFirst). */
    fun replace(
        text: String?,
        pattern: String?,
        replacement: String?,
    ): String = replaceFirst(text, pattern, replacement)

    /** Replace match đầu tiên theo pattern. */
    fun replaceFirst(
        text: String?,
        pattern: String?,
        replacement: String?,
    ): String {
        if (text == null || pattern == null) return text ?: ""
        val m = compiled(pattern) ?: return text
        return m.replaceFirst(text, replacement ?: "")
    }

    /** Replace tất cả occurrences theo pattern. */
    fun replaceAll(
        text: String?,
        pattern: String?,
        replacement: String?,
    ): String {
        if (text == null || pattern == null) return text ?: ""
        val m = compiled(pattern) ?: return text
        return m.replace(text, replacement ?: "")
    }

    /** Split text theo pattern (limit -1, giữ trailing empty). */
    fun split(
        text: String?,
        pattern: String?,
    ): Array<String> {
        if (text == null || pattern == null) {
            return if (text != null) arrayOf(text) else arrayOf()
        }
        return try {
            // limit lớn dương = giữ trailing empty (Java Pattern.split limit âm semantics —
            // Kotlin stdlib split không nhận limit âm)
            Regex(pattern).split(text, Int.MAX_VALUE).toTypedArray()
        } catch (e: Exception) {
            log.add("Invalid regex pattern: $pattern")
            arrayOf(text)
        }
    }

    /** Test pattern có match không. */
    fun test(
        text: String?,
        pattern: String?,
    ): Boolean {
        if (text == null || pattern == null) return false
        val m = compiled(pattern) ?: return false
        return m.containsMatchIn(text)
    }

    /** Escape regex special characters. */
    fun escape(text: String?): String = if (text == null) "" else Regex.escape(text)

    /** Find match cuối cùng. */
    fun findLast(
        text: String?,
        pattern: String?,
    ): String {
        if (text == null || pattern == null) return ""
        val m = compiled(pattern) ?: return ""
        return m.findAll(text).lastOrNull()?.value ?: ""
    }

    /** Đếm occurrences theo pattern. */
    fun count(
        text: String?,
        pattern: String?,
    ): Int {
        if (text == null || pattern == null) return 0
        val m = compiled(pattern) ?: return 0
        return m.findAll(text).count()
    }

    /** Lấy tất cả groups từ match đầu tiên. */
    fun groups(
        text: String?,
        pattern: String?,
    ): Array<String?> {
        if (text == null || pattern == null) return arrayOf()

        val groupList = ArrayList<String?>()
        val m = compiled(pattern)
        val match = m?.find(text)
        if (match != null) {
            for (i in 0..match.groups.size - 1) groupList.add(match.groups[i]?.value)
        }
        return groupList.toTypedArray()
    }

    /** Trích groups dạng "group1", "group2", ... từ match đầu tiên. */
    fun namedGroups(
        text: String?,
        pattern: String?,
    ): MutableMap<String, String?> {
        val result = HashMap<String, String?>()
        if (text == null || pattern == null) return result

        val match = compiled(pattern)?.find(text)
        if (match != null) {
            for (i in 1..match.groups.size - 1) result["group$i"] = match.groups[i]?.value
        }
        return result
    }

    /** Check pattern hợp lệ. */
    fun isValid(pattern: String?): Boolean {
        if (pattern.isNullOrEmpty()) return false
        return try {
            Regex(pattern)
            true
        } catch (e: Exception) {
            false
        }
    }

    /** Lấy match tại index cụ thể. */
    fun matchAt(
        text: String?,
        pattern: String?,
        index: Int,
    ): String {
        if (text == null || pattern == null || index < 0) return ""

        val m = compiled(pattern) ?: return ""
        return m.findAll(text).elementAtOrNull(index)?.value ?: ""
    }

    /** Tất cả matches kèm vị trí. */
    fun matchesWithPositions(
        text: String?,
        pattern: String?,
    ): Array<HashMap<String, Any?>> {
        if (text == null || pattern == null) return arrayOf()

        val results = ArrayList<HashMap<String, Any?>>()
        val m = compiled(pattern)
        if (m != null) {
            m.findAll(text).forEach { match ->
                val m2 = HashMap<String, Any?>()
                m2["text"] = match.value
                m2["start"] = match.range.first
                m2["end"] = match.range.last + 1
                results.add(m2)
            }
        }
        return results.toTypedArray()
    }
}
